File: src/test/java/com/tngtech/jgiven/scenario/state/ScenarioStateAnnotationProviderTest.java
package com.tngtech.jgiven.scenario.state;

import com.intellij.psi.PsiAnnotation;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiField;
import com.tngtech.jgiven.Annotations;
import com.tngtech.jgiven.util.AnnotationProvider;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class ScenarioStateAnnotationProviderTest {
    @Rule
    public MockitoRule mockitoRule = MockitoJUnit.rule();

    @Mock
    private AnnotationProvider annotationProvider;

    @InjectMocks
    private ScenarioStateAnnotationProvider scenarioStateAnnotationProvider;

    @Before
    public void setUp() {
        scenarioStateAnnotationProvider = new ScenarioStateAnnotationProvider(annotationProvider);
    }

    @Test
    public void determine_JGiven_ScenarioState() {
        PsiField field = mock(PsiField.class);
        when(annotationProvider.findAnnotation(field, Annotations.INSTANCE.getJGIVEN_SCENARIO_STATE_CLASS_NAMES())).thenReturn(mock(PsiAnnotation.class));

        boolean result = scenarioStateAnnotationProvider.isJGivenScenarioState(field);

        assertThat(result).isTrue();
    }

    @Test
    public void determine_JGiven_if_element_is_not_a_field() {
        PsiElement element = mock(PsiElement.class);

        boolean result = scenarioStateAnnotationProvider.isJGivenScenarioState(element);

        assertThat(result).isFalse();
    }

    @Test
    public void determine_JGiven_if_element_is_not_a_scenario_state() {
        PsiField field = mock(PsiField.class);
        when(annotationProvider.findAnnotation(field, Annotations.INSTANCE.getJGIVEN_SCENARIO_STATE_CLASS_NAMES())).thenReturn(null);

        boolean result = scenarioStateAnnotationProvider.isJGivenScenarioState(field);

        assertThat(result).isFalse();
    }
}

File: src/test/java/com/tngtech/jgiven/scenario/state/ScenarioStateAnnotationProviderTest.kt
package com.tngtech.jgiven.scenario.state

import com.intellij.psi.PsiAnnotation
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiField
import com.tngtech.jgiven.Annotations
import com.tngtech.jgiven.util.AnnotationProvider
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import org.mockito.junit.MockitoJUnit
import org.mockito.junit.MockitoRule

class ScenarioStateAnnotationProviderTest {
    @Rule
    @JvmField
    var mockitoRule: MockitoRule = MockitoJUnit.rule()

    @Mock
    private lateinit var annotationProvider: AnnotationProvider

    private lateinit var scenarioStateAnnotationProvider: ScenarioStateAnnotationProvider

    @Before
    fun setUp() {
        scenarioStateAnnotationProvider = ScenarioStateAnnotationProvider(annotationProvider)
    }

    @Test
    fun determine_JGiven_ScenarioState() {
        val field = mock(PsiField::class.java)
        `when`<PsiAnnotation>(annotationProvider.findAnnotation(field, Annotations.JGIVEN_SCENARIO_STATE_CLASS_NAMES)).thenReturn(mock(PsiAnnotation::class.java))

        val result = scenarioStateAnnotationProvider.isJGivenScenarioState(field)

        assertThat(result).isTrue()
    }

    @Test
    fun determine_JGiven_if_element_is_not_a_field() {
        val element = mock(PsiElement::class.java)

        val result = scenarioStateAnnotationProvider.isJGivenScenarioState(element)

        assertThat(result).isFalse()
    }

    @Test
    fun determine_JGiven_if_element_is_not_a_scenario_state() {
        val field = mock(PsiField::class.java)
        `when`<PsiAnnotation>(annotationProvider.findAnnotation(field, Annotations.JGIVEN_SCENARIO_STATE_CLASS_NAMES)).thenReturn(null)

        val result = scenarioStateAnnotationProvider.isJGivenScenarioState(field)

        assertThat(result).isFalse()
    }
}
